build/circle-test.sh

```diff
@@ -0,0 +1,71 @@
+#!/bin/bash
+
+set -e
+
+outdir="${TMPDIR}"
+if [ "${CIRCLE_ARTIFACTS}" != "" ]; then
+    outdir="${CIRCLE_ARTIFACTS}"
+fi
+
+builder=$(dirname $0)/builder.sh
+
+# 1. Run "make check" to verify coding guidelines.
+echo "make check"
+time ${builder} make GITHOOKS= check | tee "${outdir}/check.log"; test ${PIPESTATUS[0]} -eq 0
+
+# 2. Verify that "go generate" was run.
+echo "verifying generated files"
+time ${builder} /bin/bash -c "(go generate ./... && git ls-files --modified --deleted --others --exclude-standard | diff /dev/null -) || (git add -A && git diff -u HEAD && false)" | tee "${outdir}/generate.log"; test ${PIPESTATUS[0]} -eq 0
+
+# 3. Run "make testrace".
+match='^panic|^[Gg]oroutine \d+|(read|write) by.*goroutine|DATA RACE'
+echo "make testrace"
+time ${builder} make GITHOOKS= testrace \
```

Comment:
I would think that if any of our tests fail we want to just stop there. I thought that is the way `circle.yml` worked. I could have sworn that when `make check` failed nothing else was run.

Regardless, I'm changing this so that it remembers the `make test` and `make testrace` status and exits with 1 if either are non-zero.